notifier/notifier.go

```diff
@@ -221,43 +246,110 @@ func (n *Notifier) collectRogueCallbacks() {
 	}
 }
 
-// Notify posts callback info to j.CallbackURL
-func (n *Notifier) Notify(j *job.Job) error {
-	j.CallbackCount++
+// MonitorDeliveries monitors the delivery reports channel of a backend.
+func (n *Notifier) MonitorDeliveries(ctx context.Context, id string, dwg *sync.WaitGroup) {
+	defer dwg.Done()
 
-	err := n.markCbInProgress(j)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case cbInfo, ok := <-n.backends[id].DeliveryReports():
+			if !ok {
+				return
+			}
+
+			err := n.handleCallbackInfo(cbInfo)
+			if err != nil {
+				n.Log.Printf("Error occured during handling callback info. Operation returned error %s", err)
+			}
+		}
+	}
+}
+
+// handleCallbackInfo handles a callback object.
+// If the callback is successful we increment stats counters and remove the job from storage.
+// For kafka if the callback is not successful we immediately mark the callback as failed since
+// kafka has already tried to retransmit a message for an amount of time. For http if the
+// callback is not successful we have already handled this case in Notify() method with retryOrFail().
+// Also not that when on http we call QueueJobForDeletion which schedules file deletion associated to our job,
+// because we are sure the other end has downloaded the file. But kakfa only ensures that a message has been
+// persisted, but most likely consumer has not recieved the message yet alone the file.
+func (n *Notifier) handleCallbackInfo(cbInfo job.Callback) error {
+	var e string
+
+	j, err := n.Storage.GetJob(cbInfo.JobID)
 	if err != nil {
-		return err
+		e = fmt.Sprintf("\nError: Could not get job %s. Operation returned error: %s", cbInfo.JobID, err)
+		return errors.New(e)
+	}
+
+	if cbInfo.Delivered {
+		cbType, _ := n.getCallbackTypeAndDestination(&j)
+
+		if cbType == HTTP {
+			err := n.Storage.QueueJobForDeletion(cbInfo.JobID)
```

Comment:
This logic is only applied if the recipient replies with HTTPStatusAccepted (201).
We should not delete all the ack'd http callbacks.

Reply:
We should discuss about this. Maybe we should drop this line and handle file deletion differently.